File: backend/routers/heists.py
"""
Heist API endpoints
"""
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.dependencies import get_current_user
from backend.models import User
from backend.schemas import HeistCreate, HeistResponse
from backend.heist_service import (
    create_heist,
    list_active_heists,
    list_archive_heists,
    list_my_heists,
    get_heist,
    abort_heist
)

router = APIRouter(prefix="/heists", tags=["heists"])


@router.get("", response_model=list[HeistResponse])
def get_active_heists(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    List all active heists (War Room)

    Returns heists with:
    - status = Active
    - deadline > now()

    Requires authentication.
    """
    return list_active_heists(db)


@router.post("", response_model=HeistResponse, status_code=status.HTTP_201_CREATED)
def create_new_heist(
    heist_data: HeistCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Create a new heist

    - **title**: Mission name
    - **target**: Target of the heist
    - **difficulty**: Training | Easy | Medium | Hard | Legendary
    - **assignee_username**: Username to assign this heist to
    - **deadline**: Deadline (must be in the future)
    - **description**: Optional mission details

    Requires authentication. Creator is set to the authenticated user.
    """
    return create_heist(db, heist_data, current_user)


@router.get("/archive", response_model=list[HeistResponse])
def get_archive_heists(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    List archived heists (Mission Archive)

    Returns heists with status = Expired or Aborted.

    Requires authentication.
    """
    return list_archive_heists(db)


@router.get("/mine", response_model=list[HeistResponse])
def get_my_heists(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    List heists created by the current user

    Returns only heists where creator_username equals the authenticated user's username.

    Requires authentication.
    """
    return list_my_heists(db, current_user)


@router.get("/{heist_id}", response_model=HeistResponse)
def get_heist_details(
    heist_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Get details of a specific heist

    Returns heist data by ID.

    Requires authentication.
    """
    return get_heist(db, heist_id)


@router.patch("/{heist_id}/abort", response_model=HeistResponse)
def abort_heist_by_id(
    heist_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Abort a heist

    Changes heist status from Active to Aborted.

    - Only the creator can abort their heist (403 if not creator)
    - Only Active heists can be aborted (409 if not Active)

    Requires authentication.
    """
    return abort_heist(db, heist_id, current_user)

File: frontend-react/src/services/heists.js
import api from './api';

/**
 * Heist Management Service
 * Handles all heist-related API calls
 */

/**
 * List all active heists (War Room)
 * @returns {Promise<Object>} List of active heists
 */
export const listActiveHeists = async () => {
  try {
    const response = await api.get('/heists');
    return { success: true, data: response.data };
  } catch (error) {
    return {
      success: false,
      error: error.response?.data?.detail || 'Failed to fetch active heists',
      data: [],
    };
  }
};

/**
 * List heists created by current user (My Assignments)
 * @returns {Promise<Object>} List of user's heists
 */
export const listMyHeists = async () => {
  try {
    const response = await api.get('/heists/mine');
    return { success: true, data: response.data };
  } catch (error) {
    return {
      success: false,
      error: error.response?.data?.detail || 'Failed to fetch your heists',
      data: [],
    };
  }
};

/**
 * List archived heists (Expired/Aborted)
 * @returns {Promise<Object>} List of archived heists
 */
export const listArchiveHeists = async () => {
  try {
    const response = await api.get('/heists/archive');
    return { success: true, data: response.data };
  } catch (error) {
    return {
      success: false,
      error: error.response?.data?.detail || 'Failed to fetch archive',
      data: [],
    };
  }
};

/**
 * Get heist by ID
 * @param {number} id - Heist ID
 * @returns {Promise<Object>} Heist details
 */
export const getHeist = async (id) => {
  try {
    const response = await api.get(`/heists/${id}`);
    return { success: true, data: response.data };
  } catch (error) {
    return {
      success: false,
      error: error.response?.data?.detail || 'Heist not found',
      data: null,
    };
  }
};

/**
 * Create a new heist
 * @param {Object} heistData - Heist data
 * @param {string} heistData.title - Mission name
 * @param {string} heistData.target - Target sector
 * @param {string} heistData.difficulty - Difficulty level
 * @param {string} heistData.assignee_username - Operative to assign
 * @param {string} heistData.deadline - ISO datetime string
 * @param {string} [heistData.description] - Optional mission details
 * @returns {Promise<Object>} Created heist
 */
export const createHeist = async (heistData) => {
  try {
    const response = await api.post('/heists', heistData);
    return { success: true, data: response.data };
  } catch (error) {
    // Handle validation errors (422)
    if (error.response?.status === 422) {
      const validationErrors = error.response.data?.detail;
      return {
        success: false,
        error: 'Blueprint incomplete. Fill all required fields.',
        validationErrors,
      };
    }

    return {
      success: false,
      error: error.response?.data?.detail || 'Failed to create heist',
    };
  }
};

/**
 * Abort a heist (creator only)
 * @param {number} id - Heist ID
 * @returns {Promise<Object>} Updated heist
 */
export const abortHeist = async (id) => {
  try {
    const response = await api.patch(`/heists/${id}/abort`);
    return { success: true, data: response.data };
  } catch (error) {
    // Handle specific error codes
    const status = error.response?.status;
    let errorMessage = 'Failed to abort heist';

    if (status === 403) {
      errorMessage = 'Only the creator can abort this mission';
    } else if (status === 404) {
      errorMessage = 'Heist not found';
    } else if (status === 409) {
      errorMessage = 'Heist is already expired or aborted';
    }

    return {
      success: false,
      error: error.response?.data?.detail || errorMessage,
    };
  }
};

export default {
  listActiveHeists,
  listMyHeists,
  listArchiveHeists,
  getHeist,
  createHeist,
  abortHeist,
};
